src/server/RHData.py

```diff
@@ -0,0 +1,840 @@
+            })
+
+        logger.info('Pilot added: Pilot {0}'.format(new_pilot.id))
+
+        return new_pilot
+
+    def alter_pilot(self, data):
+        pilot_id = data['pilot_id']
+        pilot = self.get_pilot(pilot_id)
+        if 'callsign' in data:
+            pilot.callsign = data['callsign']
+        if 'team_name' in data:
+            pilot.team = data['team_name']
+        if 'phonetic' in data:
+            pilot.phonetic = data['phonetic']
+        if 'name' in data:
+            pilot.name = data['name']
+
+        self._Database.DB.session.commit()
+
+        self._RACE.cacheStatus = Results.CacheStatus.INVALID  # refresh current leaderboard
+
+        self._Events.trigger(Evt.PILOT_ALTER, {
+            'pilot_id': pilot_id,
+            })
+
+        logger.info('Altered pilot {0} to {1}'.format(pilot_id, data))
+
+        race_list = []
+        if 'callsign' in data or 'team_name' in data:
+            heatnodes = self.get_heatNodes(filter_by={"pilot_id": pilot_id})
+            if heatnodes:
+                for heatnode in heatnodes:
+                    heat = self.get_heat(heatnode.heat_id)
+                    heat.cacheStatus = Results.CacheStatus.INVALID
+                    if heat.class_id != RHUtils.CLASS_ID_NONE:
+                        race_class = self.get_raceClass(heat.class_id)
+                        race_class.cacheStatus = Results.CacheStatus.INVALID
+                    for race in self.get_savedRaceMetas(
+                        filter_by={"heat_id": heatnode.heat_id}
+                        ):
+                        race_list.append(race)
+
+            if len(race_list):
+                self._PageCache.set_valid(False)
+                self.set_option("eventResults_cacheStatus", Results.CacheStatus.INVALID)
+
+                for race in race_list:
+                    race.cacheStatus = Results.CacheStatus.INVALID
+
+                self._Database.DB.session.commit()
+
+        return pilot, race_list
+
+    def delete_pilot(self, pilot_id):
+        pilot = self.get_pilot(pilot_id)
+
+        has_race = self.get_savedPilotRaces(
+            filter_by={"pilot_id": pilot.id},
+            return_type='first'
+            )
+
+        if has_race:
+            logger.info('Refusing to delete pilot {0}: is in use'.format(pilot.id))
+            return False
+        else:
+            self._Database.DB.session.delete(pilot)
+            for heatNode in self.get_heatNodes():
+                if heatNode.pilot_id == pilot.id:
+                    heatNode.pilot_id = RHUtils.PILOT_ID_NONE
+            self._Database.DB.session.commit()
+
+            logger.info('Pilot {0} deleted'.format(pilot.id))
+
+            self._RACE.cacheStatus = Results.CacheStatus.INVALID  # refresh leaderboard
+
+            return True
+
+    # Heats
+    def get_heat(self, heat_id):
+        return self._Database.Heat.query.get(heat_id)
+
+    @_Decorators.getter_parameters
+    def get_heats(self, **kwargs):
```

Comment:
get_heatNode_by_heat, etc which I think was in a previous commit was better as it hides the querying. By exposing the querying, you can end up with multiple different queries in different places that functionally do the same thing. The other thing we want to do is potentially avoid or be careful about returning db objects, as these can expose db specific fields, i.e. PKs and FKs used for relationships. Atm, if we wanted to alter the table structure, it would be a nightmare as we would have to hunt down all queries, and a lack of tests. With a fully encapsulated RHData, there will be one place to alter, one place to test.

*But* we don't need to get there in one big bang. Make *an* improvement, get it in, then make another, etc. Else we just going to end up in a mess like with my Multifrequency+stm32 PR. Merge early, merge frequently (and preferably include tests). Big bang development is old-fashioned, and I think you've seen enough of the problems it can cause first-hand. The holy-grail here is CI/CD, continuous integration, continuous delivery, continuous deployment (e.g. Pawel's OTA is an enabler for this).

Reply:
I considered converting the results to a dict or object type before returning them so they are unlinked from the DB but I can't be sure yet that some of the code isn't doing that already. Working toward that point. The builds didn't run on the last commit for whatever reason, but as of the latest the existing tests were passing.

Are you saying that what I have is in a good enough place to roll in?

Reply:
Yes, because it is much better than what it was. For example, do the get methods make it easy to mock? No, but it is now a more feasible proposition. It is more achievable than it was.

I think the builds didnt run because we ran out of credits. From what I can tell, we might be caught in a backlog, and probably have to migrate to github actions.